### ISSUE TITLE
Tests failed when using Python 3.12.1

### ISSUE THREAD

**Issue by tklam** [magia-hdl/magia#19]
Using Python 3.12.1 leads to many failed test cases:
```console
poetry shell
poetry env use python3.12
poetry install --with=dev
poetry run pytest 
```
Then the test results (attached)
[magia.test.19jan2024.txt](https://github.com/magia-hdl/magia/files/13979122/magia.test.19jan2024.txt)
:
```
============================= test session starts ==============================
platform linux -- Python 3.12.1, pytest-7.4.4, pluggy-1.3.0
rootdir: /home/tklam/development/magia-hdl/magia
configfile: pyproject.toml
plugins: cocotb-test-0.2.4, xdist-3.5.0, github-report-0.0.1
collected 186 items
 
tests/test_external_mod.py .................                             [  9%] 
tests/test_memory.py FFFF                                                [ 11%] 
tests/test_module_instance.py .....                                      [ 13%]
tests/test_smoke.py ...                                                  [ 15%]
tests/bundle/test_bundle.py FF                                           [ 16%]
tests/core/test_const.py F........                                       [ 21%]
tests/core/test_op.py FF........FFFFFFFFFFFFF                            [ 33%]
tests/core/test_reg.py FFFFFFFFF                                         [ 38%]
tests/core/test_signal.py .F.                                            [ 40%]
tests/std/test_fixed_point.py .......................................... [ 62%] 
....................................................                     [ 90%] 
tests/std/test_stdio.py F                                                [ 91%] 
tests/std/encoding/test_onehot.py FFFFF.FFFFFFFFFF                       [100%] 

```

**Comment by khwong-c** [comment]
## TL;DR
It's a problem of Verilator. It used the System Default `/usr/bin/python3` as it's python interpreter internally.

**Comment by khwong-c** [comment]
It is mentioned an override on PYTHON3 In Verilator's Document.
However, it is not respected in runtime, on 5.020 the latest version.

It can be changed only when compiling `verilator` itself. 
This will add an extra 5min to our CI, and this is expensive.
Furthermore, this doesn't fix our situation when individual developer try to integrate their dev environment with verilator.

I couldn't think of a decent way to fix that.
For the CI, the quick and dirty way is "Relink `/usr/bin/python3` with the `python` in the virtual env"

Yup. It is really dirty.


https://verilator.org/guide/latest/environment.html
![image](https://github.com/magia-hdl/magia/assets/92428218/95ff1164-be54-4c93-873c-b47f05ea77ac)

**Comment by khwong-c** [comment]
Fixed the Issue on CI with #16 
But we need to handle this later on when we integrate `magia-hdl` with the simulation workflow.

@tklam proposed to raise an issue/PR on verilator's repo.
